In [29]:
import pandas as pd
import os
from pathlib import Path
import matplotlib.pyplot as plt
import seaborn as sns
from collections import Counter
import numpy as np

In [3]:
dataset_path = os.path.join(Path.cwd().parent, 'data', 'stackexchange_dataset.csv')
data = pd.read_csv(dataset_path,index_col='question_id')

In [4]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Index: 100000 entries, 79802517 to 17945108
Data columns (total 11 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   title                          100000 non-null  object 
 1   has_accepted_answer            100000 non-null  bool   
 2   accepted_answer_score          12000 non-null   float64
 3   time_to_accepted_answer_hours  12000 non-null   float64
 4   question_score                 100000 non-null  int64  
 5   question_text                  100000 non-null  object 
 6   num_tags                       100000 non-null  int64  
 7   tags                           100000 non-null  object 
 8   accepted_answer_id             12000 non-null   float64
 9   accepted_answer_length_chars   12000 non-null   float64
 10  accepted_answer_length_tokens  12000 non-null   float64
dtypes: bool(1), float64(5), int64(2), object(3)
memory usage: 8.5+ MB


In [5]:
data.head()

title  \
question_id                                                      
79802517     Looking for a better way using &quot;.Include&...   
79802934     NTP is moving my clock further from the correc...   
79802909               Execution of pandas&#39; info in python   
79802932     How to debug MongoDB recurring error 314 Objec...   
79802919     How to integrate QML UI into a custom Vulkan r...   

             has_accepted_answer  accepted_answer_score  \
question_id                                               
79802517                   False                    NaN   
79802934                   False                    NaN   
79802909                   False                    NaN   
79802932                   False                    NaN   
79802919                   False                    NaN   

             time_to_accepted_answer_hours  question_score  \
question_id                                                  
79802517                               NaN               2   
79802934                               NaN               0   
79802909                               NaN               0   
79802932                               NaN               0   
79802919                               NaN               0   

                                                 question_text  num_tags  \
question_id                                                                
79802517     I am looking for a better way to use the .Incl...         2   
79802934     Windows 11 Pro 10.0.26200 Build 26200. Dell XP...         1   
79802909     I am new to pandas library in python. When I l...         2   
79802932     On a fresh Mongo DB 8.0 mono instance (no clus...         4   
79802919     I'm developing a custom Vulkan renderer and wa...         4   

                                                          tags  \
question_id                                                      
79802517                            ['c#', 'entity-framework']   
79802934                                               ['ntp']   
79802909                                  ['python', 'pandas']   
79802932     ['database', 'mongodb', 'nosql', 'system-admin...   
79802919                        ['c++', 'qt', 'qml', 'vulkan']   

             accepted_answer_id  accepted_answer_length_chars  \
question_id                                                     
79802517                    NaN                           NaN   
79802934                    NaN                           NaN   
79802909                    NaN                           NaN   
79802932                    NaN                           NaN   
79802919                    NaN                           NaN   

             accepted_answer_length_tokens  
question_id                                 
79802517                               NaN  
79802934                               NaN  
79802909                               NaN  
79802932                               NaN  
79802919                               NaN

In [6]:
data.index.is_unique

False

In [7]:
data[data.index.duplicated()].shape[0]

8

In [8]:
data[data.index.duplicated(keep=False)].head(4)

title  \
question_id                                                      
79802647                   Paypal webhook simulator not firing   
79802647                   Paypal webhook simulator not firing   
79789014     Can I use SSM to run command for SQL query exe...   
79789014     Can I use SSM to run command for SQL query exe...   

             has_accepted_answer  accepted_answer_score  \
question_id                                               
79802647                   False                    NaN   
79802647                   False                    NaN   
79789014                   False                    NaN   
79789014                   False                    NaN   

             time_to_accepted_answer_hours  question_score  \
question_id                                                  
79802647                               NaN               0   
79802647                               NaN               0   
79789014                               NaN              -1   
79789014                               NaN              -1   

                                                 question_text  num_tags  \
question_id                                                                
79802647     To test a PayPal integration that also reacts ...         2   
79802647     To test a PayPal integration that also reacts ...         2   
79789014     I want to automate a database drop and databas...         3   
79789014     I want to automate a database drop and databas...         3   

                                                         tags  \
question_id                                                     
79802647                               ['paypal', 'webhooks']   
79802647                               ['paypal', 'webhooks']   
79789014     ['sql-server', 'amazon-web-services', 'aws-ssm']   
79789014     ['sql-server', 'amazon-web-services', 'aws-ssm']   

             accepted_answer_id  accepted_answer_length_chars  \
question_id                                                     
79802647                    NaN                           NaN   
79802647                    NaN                           NaN   
79789014                    NaN                           NaN   
79789014                    NaN                           NaN   

             accepted_answer_length_tokens  
question_id                                 
79802647                               NaN  
79802647                               NaN  
79789014                               NaN  
79789014                               NaN

Since they are the same question, we drop them

In [9]:
data2 = data[~data.index.duplicated()]

In [10]:
data2.shape,data.shape

((99992, 11), (100000, 11))

In [11]:
data2.index.is_unique

True

---

In [12]:
data2['question_score'].max(),data2['question_score'].min()

(np.int64(27487), np.int64(-20))

In [13]:
data2['has_accepted_answer'].where(data2['has_accepted_answer']==True).dropna().count(),data2['has_accepted_answer'].where(data2['has_accepted_answer']==False).dropna().count()

(np.int64(39938), np.int64(60054))

Huge disproportion in the data in regards to accepted answer

In [14]:
data_acc = data2[data2['has_accepted_answer']]

In [15]:
data_acc.shape

(39938, 11)

In [16]:
data_acc[data_acc['time_to_accepted_answer_hours'].notna()].shape

(12000, 11)

Out of those 39938 questions, only 12000 has accepted answer

In [17]:
data_aac = data_acc[data_acc['time_to_accepted_answer_hours'].notna()]
data_aac.shape

(12000, 11)

In [19]:
data_aac

title  \
question_id                                                      
1562367                       How do short URLs services work?   
79802633     How to import based on variable to avoid loadi...   
79801576     Is std::unique_ptr needed on abstract class wh...   
46853200     How to make zsh&#39;s REPORTTIME work? (time f...   
2711001      How to apply shell command to each line of a c...   
...                                                        ...   
25045932                             Fake DbFunctions DiffDays   
71355883                 Next-Auth Signout within jwt callback   
78467875     freeing allocated memory returned from functio...   
79662126     Is there a built-in identity function in JavaS...   
79663172                    How to get a clickable PDF in HTML   

             has_accepted_answer  accepted_answer_score  \
question_id                                               
1562367                     True                  242.0   
79802633                    True                    1.0   
79801576                    True                    1.0   
46853200                    True                    8.0   
2711001                     True                  347.0   
...                          ...                    ...   
25045932                    True                    0.0   
71355883                    True                   20.0   
78467875                    True                   -1.0   
79662126                    True                    5.0   
79663172                    True                    0.0   

             time_to_accepted_answer_hours  question_score  \
question_id                                                  
1562367                           0.499167             119   
79802633                          2.591667               0   
79801576                         10.045833               1   
46853200                          2.211667               6   
2711001                           0.055000             328   
...                                    ...             ...   
25045932                       6541.563889               1   
71355883                        481.200556              10   
78467875                          0.244444              -1   
79662126                          0.059167               0   
79663172                          0.138889              -3   

                                                 question_text  num_tags  \
question_id                                                                
1562367      How do services like TinyURL or Metamark work?...         5   
79802633     Using React Native+Expo I'm trying to load onl...         5   
79801576     I came across below code in a project I am wor...         5   
46853200     This is my .zshrc: export REPORTTIME=3 When I ...         1   
2711001      Suppose I have some output from a command (suc...         1   
...                                                        ...       ...   
25045932     I am trying to fake DbFunctions.DiffDays but a...         2   
71355883     If a user is signed into the app between app u...         4   
78467875     If I have a function like this: char* f1(char ...         3   
79662126     Is there a function in JavaScript that accepts...         1   
79663172     To have a picture (eg. in jpg format) on a web...         2   

                                                          tags  \
question_id                                                      
1562367      ['web-services', 'url', 'url-shortener', 'bit....   
79802633     ['javascript', 'reactjs', 'node.js', 'typescri...   
79801576     ['c++', 'c++20', 'unique-ptr', 'std-function',...   
46853200                                               ['zsh']   
2711001                                               ['bash']   
...                                                        ...   
25045932                                ['c#', 'unit-testing']   
71355883     ['javascript', 'reactjs', '

---
## Tags

In [ ]:
import ast

In [65]:
tags=data_aac[['tags','num_tags']]
tags.loc[:,('tags')] = tags.loc[:,('tags')].apply(ast.literal_eval)


In [66]:
tags.iloc[0]['tags']

['web-services', 'url', 'url-shortener', 'bit.ly', 'short-url']

In [75]:
tag_counter = Counter(tag for tags in tags['tags'] for tag in tags)
tag_freq = pd.Series(tag_counter).sort_values(ascending=False)

In [76]:
tag_freq

python             1528
c#                  746
javascript          703
c++                 689
java                592
                   ... 
gocql                 1
screen                1
b-tree-index          1
beeware               1
outlook-restapi       1
Length: 7684, dtype: int64

### Problem: Extreme Multi-label Classification

As seen above, there are $7684$ unique tags. This is too big for a classification problem.

We can use:
- Frequency-Based Filtering
    - Keep only the top `N` most frequent tags
- "Semantic Compression" using Embeddings and Clustering
  - Use embeddings and clustering to group them

